core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -0,0 +1,220 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Arm Limited
+ */
+#include <assert.h>
+#include <bench.h>
+#include <kernel/panic.h>
+#include <kernel/secure_partition.h>
+#include <kernel/spinlock.h>
+#include <kernel/spmc_sp_handler.h>
+#include <optee_ffa.h>
+#include "thread_private.h"
+
+void spmc_sp_start_thread(struct thread_smc_args *args)
+{
+	thread_sp_alloc_and_run(args);
+}
+
+static void ffa_set_error(struct thread_smc_args *args, uint32_t error)
+{
+	args->a0 = FFA_ERROR;
+	args->a2 = error;
+}
+
+static TEE_Result ffa_get_dst(struct thread_smc_args *args,
+			      struct sp_session *caller,
+			      struct sp_session **dst)
+{
+	if (args->a2 != FFA_PARAM_MBZ) {
+		/* a2 should always be 0 */
+		*dst = NULL;
+		return FFA_INVALID_PARAMETERS;
+	}
+
+	*dst = sp_get_session(FFA_DST(args->a1));
+
+	/* Message came from the NW */
+	if (!caller) {
+		if (!*dst) {
+			EMSG("Neither destination nor source is a SP!");
+			return FFA_INVALID_PARAMETERS;
+		}
+	} else {
+		/* Check if the source matches the endpoint we came from */
+		if (FFA_SRC(args->a1) != caller->endpoint_id) {
+			EMSG("Source address doesn't match the endpoint id!");
+			return FFA_INVALID_PARAMETERS;
+		}
+	}
+	return FFA_OK;
+}
+
+static struct sp_session *
+ffa_handle_sp_direct_req(struct thread_smc_args *args,
+			 struct sp_session *caller_sp)
+{
+	struct sp_session *dst = NULL;
+	TEE_Result res = FFA_OK;
+
+	if (args->a2 != FFA_PARAM_MBZ) {
+		/* a2 should always be 0 */
+		dst = NULL;
+		ffa_set_error(args, FFA_INVALID_PARAMETERS);
+		return NULL;
+	}
+
+	res = ffa_get_dst(args, caller_sp, &dst);
+	if (res) {
+		/* Tried to send message to an incorrect endpoint */
+		ffa_set_error(args, res);
+		return caller_sp;
+	}
+	if (!dst) {
+		EMSG("Request to normal world not supported!");
+		ffa_set_error(args, FFA_NOT_SUPPORTED);
+		return NULL;
+	}
+
+	cpu_spin_lock(&dst->spinlock);
+	if (dst->state != sp_idle) {
+		EMSG("SP is busy");
+		ffa_set_error(args, FFA_BUSY);
+		cpu_spin_unlock(&dst->spinlock);
+		return caller_sp;
+	}
+
+	dst->state = sp_busy;
+	cpu_spin_unlock(&dst->spinlock);
+
+	/*
+	 * Store the calling endpoint id. This will make it possible to check
+	 * if the response is send back to the correct endpoint.
+	 */
+	dst->caller_id = FFA_SRC(args->a1);
+
+	/* Forward the message to the destination Secure Partition. */
```

Comment:
s/Secure Partition/SP/
for consistency.

Reply:
Done